node/src/components/binary_port.rs

```diff
@@ -0,0 +1,887 @@
+                path,
+                protocol_version,
+            )
+            .await
+        }
+        GetRequest::AllValues {
+            state_root_hash,
+            key_tag,
+        } => {
+            metrics.binary_port_get_all_values_count.inc();
+            if !config.allow_request_get_all_values {
+                BinaryResponse::new_error(
+                    binary_port::ErrorCode::FunctionDisabled,
+                    protocol_version,
+                )
+            } else {
+                handle_get_all_values(state_root_hash, key_tag, effect_builder, protocol_version)
+                    .await
+            }
+        }
+        GetRequest::Trie { trie_key } => {
+            metrics.binary_port_get_trie_count.inc();
+            let response = if !config.allow_request_get_trie {
+                BinaryResponse::new_error(
+                    binary_port::ErrorCode::FunctionDisabled,
+                    protocol_version,
+                )
+            } else {
+                match effect_builder.get_trie_full(trie_key).await {
+                    Ok(result) => BinaryResponse::from_value(result, protocol_version),
+                    Err(_err) => BinaryResponse::new_error(
+                        binary_port::ErrorCode::InternalError,
+                        protocol_version,
+                    ),
+                }
+            };
+            response
+        }
+    }
+}
+
+async fn handle_get_all_values<REv>(
+    state_root_hash: Digest,
+    key_tag: casper_types::KeyTag,
+    effect_builder: EffectBuilder<REv>,
+    protocol_version: ProtocolVersion,
+) -> BinaryResponse
+where
+    REv: From<Event> + From<ContractRuntimeRequest>,
+{
+    let get_all_values_request = GetAllValuesRequest::new(state_root_hash, key_tag);
+    match effect_builder.get_all_values(get_all_values_request).await {
+        Ok(GetAllValuesResult::Success { values }) => {
+            BinaryResponse::from_value(values, protocol_version)
+        }
+        Ok(GetAllValuesResult::RootNotFound) => {
+            let error_code = binary_port::ErrorCode::RootNotFound;
+            BinaryResponse::new_error(error_code, protocol_version)
+        }
+        Err(_err) => {
+            BinaryResponse::new_error(binary_port::ErrorCode::InternalError, protocol_version)
+        }
+    }
+}
+
+async fn handle_get_state_request<REv>(
+    effect_builder: EffectBuilder<REv>,
+    state_root_hash: Digest,
+    base_key: casper_types::Key,
+    path: Vec<String>,
+    protocol_version: ProtocolVersion,
+) -> BinaryResponse
+where
+    REv: From<Event> + From<ContractRuntimeRequest>,
+{
+    match effect_builder
+        .query_global_state(QueryRequest::new(state_root_hash, base_key, path))
+        .await
+    {
+        Ok(QueryResult::Success { value, proofs }) => match proofs.to_bytes() {
+            Ok(proofs) => BinaryResponse::from_value(
+                GlobalStateQueryResult::new(*value, base16::encode_lower(&proofs)),
```

Comment:
Could we change the `GlobalStateQueryResult` to not hold the proofs as an encoded string?  Seems like we should be able to just hold them as `Vec<TrieMerkleProof<Key, StoredValue>>`.

Reply:
we've considered that, but it'd require extracting these three types into `casper-types`:
- `TrieMerkleProof`
- `TrieMerkleProofStep`
- `Pointer`
and we thought that might be pulling too much stuff into types

Reply:
methods on these types seem to expose some internal functionality:
- `pub fn proof_steps(&self) -> &VecDeque<TrieMerkleProofStep>` on `TrieMerkleProof`
- `TrieMerkleProofStep` is an enum which means it's variants will become our public API:
  - ```rust
    /// Corresponds to [`Trie::Node`]
    Node {
        /// Hole index.
        hole_index: u8,
        /// Indexed pointers with hole.
        indexed_pointers_with_hole: Vec<(u8, Pointer)>,
    },
    /// Corresponds to [`Trie::Extension`]
    Extension {
        /// Affix bytes.
        affix: Bytes,
    },
    ```

Reply:
do we want to put all of this in the public API?

Reply:
That gets my vote, yes.  I approved the PR you made to move those types.

Reply:
ok, merged it in 👍